src/couch_stats/src/csrt_server.erl

```diff
@@ -0,0 +1,198 @@
+% the License.
+
+-module(csrt_server).
+
+-behaviour(gen_server).
+
+-export([
+    start_link/0,
+    init/1,
+    handle_call/3,
+    handle_cast/2
+]).
+
+-export([
+    create_pid_ref/0,
+    create_resource/1,
+    destroy_resource/1,
+    get_resource/1,
+    get_context_type/1,
+    inc/2,
+    inc/3,
+    new_context/2,
+    set_context_dbname/2,
+    set_context_username/2,
+    set_context_type/2,
+    update_counter/3
+]).
+
+-include_lib("stdlib/include/ms_transform.hrl").
+-include_lib("couch_stats_resource_tracker.hrl").
+
+
+-record(st, {}).
+
+%%
+%% Public API
+%%
+
+-spec create_pid_ref() -> pid_ref().
+create_pid_ref() ->
+    {self(), make_ref()}.
+
+%%
+%%
+%% Context lifecycle API
+%%
+
+-spec new_context(Type :: rctx_type(), Nonce :: nonce()) -> rctx().
+new_context(Type, Nonce) ->
+    #rctx{
+       nonce = Nonce,
+       pid_ref = create_pid_ref(),
+       type = Type
+    }.
+
+-spec set_context_dbname(DbName, PidRef) -> boolean() when
+    DbName :: dbname(), PidRef :: maybe_pid_ref().
+set_context_dbname(_, undefined) ->
+    false;
+set_context_dbname(DbName, PidRef) ->
+    update_element(PidRef, [{#rctx.dbname, DbName}]).
+
+%%set_context_handler_fun(_, undefined) ->
+%%    ok;
+%%set_context_handler_fun(Fun, PidRef) when is_function(Fun) ->
+%%    FProps = erlang:fun_info(Fun),
+%%    Mod = proplists:get_value(module, FProps),
+%%    Func = proplists:get_value(name, FProps),
+%%    #rctx{type=#coordinator{}=Coordinator} = get_resource(PidRef),
+%%    Update = [{#rctx.type, Coordinator#coordinator{mod=Mod, func=Func}}],
+%%    update_element(PidRef, Update).
+
+-spec set_context_username(UserName, PidRef) -> boolean() when
+    UserName :: username(), PidRef :: maybe_pid_ref().
+set_context_username(_, undefined) ->
+    ok;
+set_context_username(UserName, PidRef) ->
+    update_element(PidRef, [{#rctx.username, UserName}]).
+
+-spec get_context_type(Rctx :: rctx()) -> rctx_type().
+get_context_type(#rctx{type=Type}) ->
+    Type.
+
+-spec set_context_type(Type, PidRef) -> boolean() when
+    Type :: rctx_type(), PidRef :: maybe_pid_ref().
+set_context_type(Type, PidRef) ->
+    update_element(PidRef, [{#rctx.type, Type}]).
+
+-spec create_resource(Rctx :: rctx()) -> true.
+create_resource(#rctx{} = Rctx) ->
+    catch ets:insert(?MODULE, Rctx).
+
+-spec destroy_resource(PidRef :: maybe_pid_ref()) -> boolean().
+destroy_resource(undefined) ->
+    false;
+destroy_resource({_,_}=PidRef) ->
+    catch ets:delete(?MODULE, PidRef).
+
+-spec get_resource(PidRef :: maybe_pid_ref()) -> maybe_rctx().
+get_resource(undefined) ->
+    undefined;
+get_resource(PidRef) ->
+    catch case ets:lookup(?MODULE, PidRef) of
+        [#rctx{}=Rctx] ->
+            Rctx;
+        [] ->
+            undefined
+    end.
+
+-spec is_rctx_field(Field :: rctx_field() | atom()) -> boolean().
+is_rctx_field(Field) ->
+    maps:is_key(Field, ?KEYS_TO_FIELDS).
+
+-spec get_rctx_field(Field :: rctx_field()) -> non_neg_integer().
```

Comment:
This function might throw an error `bad key`. So technically the type spec is wrong. However I do understand that we use macro definitions for field names so the chance of a typo is quite low. If you decide to update the spec here is what I suggest.

```
%% maybe define this type in the couch_stats_resource_tracker.hrl file
-type throw(_Reason) :: no_return().

-spec get_rctx_field(Field :: rctx_field()) -> non_neg_integer() 
    | throw({badkey, Key :: any()}).
```

Reply:
Oh that's a nice way of defining exceptions! I was really hoping someone was going to provide a solid suggestion on that front, thanks! :D